### PR TITLE
Switch to Bazel 4, migrate ObjcProvider compile info to CcCompilationContext

### PR DESCRIPTION
**Pull request by dgcoffman**
Resolves
https://github.com/pinterest/PodToBUILD/issues/170
https://github.com/pinterest/PodToBUILD/issues/171
https://github.com/pinterest/PodToBUILD/issues/173

### CODE REVIEW

File: .bazelrc [hunk 1]
```diff
@@ -1,11 +1,11 @@
...
 build \
- --spawn_strategy=standalone \
+ --spawn_strategy=local \
  --strategy=SwiftCompile=worker \
  --swiftcopt=-Xwrapped-swift=-debug-prefix-pwd-is-dot \
  --experimental_strict_action_env=true
 
 test \
- --spawn_strategy=standalone \
+ --spawn_strategy=local \
```
Comment (dgcoffman) [inline .bazelrc:10]: spawn_strategy `standalone` is deprecated. Docs say "Please use local instead."

https://docs.bazel.build/versions/master/user-manual.html#flag--spawn_strategy

File: .gitignore [hunk 2]
```diff
@@ -15,3 +15,4 @@ external
 Pods/
 PodsHost/
 *.xcworkspace
+PodToBUILD.zip
```
Comment (dgcoffman) [inline .gitignore:18]: This doesn't seem to be checked in, but is produces by e.g. `make build-test`

File: BUILD [hunk 3]
```diff
@@ -1,5 +1,6 @@
 load("@build_bazel_rules_swift//swift:swift.bzl", "swift_library")
 load("@build_bazel_rules_apple//apple:macos.bzl", "macos_command_line_application", "macos_unit_test")
+load("@rules_cc//cc:defs.bzl", "objc_library")
```
Comment (dgcoffman) [inline BUILD:3]: Buildifier now warns

> Function "objc_library" is not global anymore and needs to be loaded from "@rules_cc//cc:defs.bzl".buildifier(native-cc)

File: BUILD [hunk 4]
```diff
@@ -8,7 +9,7 @@ objc_library(
     includes = ["Sources/ObjcSupport/include"]
 )
 
-# PodToBUILD is a core library enabling Skylark code generation
+# PodToBUILD is a core library enabling Starlark code generation
```
Comment (dgcoffman) [inline BUILD:12]: Skylark was renamed to "Starlark" in August 2018 https://blog.bazel.build/2018/08/17/starlark.html

File: BazelExtensions/extensions.bzl [hunk 5]
```diff
@@ -1,5 +1,5 @@
 # This file is part of PodSpecToBUILD
-# Warning: this file is not accounted for as an explict imput into the build.
+# Warning: this file is not accounted for as an explicit input into the build.
```
Comment (dgcoffman) [inline BazelExtensions/extensions.bzl:2]: "imput" typo

File: BazelExtensions/extensions.bzl [hunk 6]
```diff
@@ -190,24 +190,31 @@ module {module_name}.Swift {{
...
     # module map is added to `objc_library` as a dep, bazel will add these
     # automatically and add a _single_ include to this module map. Ideally there
-    # would be an API to invoke clang with -fmodule-map= 
+    # would be an API to invoke clang with -fmodule-map=
+    providers = []
     if ctx.attr.module_map_name == "module.modulemap":
         provider_hdr = [module_map] + ([umbrella_header_file] if umbrella_header_file else [])
         objc_provider = apple_common.new_objc_provider(
             module_map=depset([module_map]),
-            include=depset([ctx.outputs.module_map.dirname]),
```
Comment (dgcoffman) [inline BazelExtensions/extensions.bzl:198]: ObjcProvider now longer takes `include`.  https://github.com/bazelbuild/bazel/issues/10674

File: Examples/ArcSplitting/WORKSPACE [hunk 7]
```diff
@@ -1,29 +1,29 @@
...
     urls = ["https://note-this-is-overridden"],
 )
 
-git_repository(
+http_archive(
     name = "build_bazel_rules_apple",
-    remote = "https://github.com/bazelbuild/rules_apple.git",
-    commit = "1cdaf74e44c4c969d7ee739b3a0f11b993c49d2a",
+    sha256 = "a41a75c291c69676b9974458ceee09aea60cee0e1ee282e27cdc90b679dfd30f",
+    url = "https://github.com/bazelbuild/rules_apple/releases/download/0.21.2/rules_apple.0.21.2.tar.gz",
```
Comment (dgcoffman) [inline Examples/ArcSplitting/WORKSPACE:11]: Upgrading to the latest release of rules_apple.

File: Examples/PINRemoteImage/Pods.WORKSPACE [hunk 8]
```diff
@@ -12,13 +12,13 @@ new_pod_repository(
...
   name = "PINOperation",
-  url = "https://github.com/pinterest/PINOperation/archive/1.1.zip",
+  url = "https://github.com/pinterest/PINOperation/archive/1.2.1.zip",
   generate_header_map = True
 )
 
 new_pod_repository(
   name = "PINCache",
-  url = "https://github.com/pinterest/PINCache/archive/d886490de6d297e38f80bb750ff2dec4822fb870.zip",
+  url = "https://github.com/pinterest/PINCache/archive/3.0.3.zip",
```
Comment (dgcoffman) [inline Examples/PINRemoteImage/Pods.WORKSPACE:21]: Fixing https://github.com/pinterest/PodToBUILD/issues/173

The older version of PINCache does not compile in Xcode 12.

File: BazelExtensions/extensions.bzl [hunk 9]
```diff
@@ -257,12 +264,22 @@ def gen_module_map(name,
...
     for target in ctx.attr.include_files:
         for f in target.files.to_list():
             includes.append(f.path)
 
-    return apple_common.new_objc_provider(
-            include=depset(includes))
+    compilation_context = cc_common.create_compilation_context(
+            includes=depset(includes))
+
+    providers = []
```
Comment (jerrymarino) [inline BazelExtensions/extensions.bzl:275]: Minor/nit - could simplify to just return the array.
```
return [
]
```

File: Examples/ChildPodspec/WORKSPACE [hunk 10]
```diff
@@ -1,29 +1,29 @@
...
-    commit = "d07d880dcf939e0ad98df4dd723f8516bf8a2867",
+apple_rules_dependencies()
+
+load(
+    "@build_bazel_apple_support//lib:repositories.bzl",
+    "apple_support_dependencies",
 )
 
-apple_rules_dependencies()
+apple_support_dependencies()
```
Comment (jerrymarino) [inline Examples/ChildPodspec/WORKSPACE:26]: So this automatically pull in `rules_swift` and `rules_cc` now - nice

File: Makefile [hunk 11]
```diff
@@ -1,8 +1,9 @@
-.PHONY: build 
+.PHONY: build
 build:
 	@tools/bazel build \
 		--disk_cache=$(HOME)/Library/Caches/Bazel \
-		--spawn_strategy=standalone \
+		--spawn_strategy=local \
+		--use_top_level_targets_for_symlinks \
```
Comment (jerrymarino) [inline Makefile:6]: @dgcoffman is this now required for the non header map build?

File: BazelExtensions/extensions.bzl [hunk 12]
```diff
@@ -313,7 +330,8 @@ def _make_headermap_impl(ctx):
         if not hasattr(hdr_provider, "objc"):
             continue
 
-        hdrs = hdr_provider.objc.header.to_list()
+        hdrs = hdr_provider.objc.direct_headers
```
Comment (jerrymarino) [inline BazelExtensions/extensions.bzl:333]: Can you confirm if this has the same impact as it previously did? I believe here it merges transitive header maps in order to propagate the transitive includes, but would need to look more into that. If it still builds the top 50 swift cocoa pods and passes tests it _should_ work.

**Review by jerrymarino — APPROVED** [review]
@dgcoffman - thank you kindly for the nice PR and updating it to work with 4.0 🥳  Overall this LGTM, my only question was around the top level symlinks and transitive hmap merging

File: Makefile (hunk 11 above)
Comment (dgcoffman) [reply Makefile:6]: Sorry, I didn't understand the question. What's the "non header map build"?

I added `--use_top_level_targets_for_symlinks` here because something (I'm not sure what but I think the updated rules_apple?) causes an additional bindir to be created, which bazel-bin then symlinks to instead of the one which actually contains RepoTools and Compiler.  `use_top_level_targets_for_symlinks` just makes bazel-bin point at the directory it did previously.

File: BazelExtensions/extensions.bzl (hunk 12 above)
Comment (dgcoffman) [reply BazelExtensions/extensions.bzl:333]: Good call, looking into it.

Comment (dgcoffman) [reply BazelExtensions/extensions.bzl:330]: from the migration description in https://github.com/bazelbuild/bazel/issues/10674, it sounds like the right thing would be to change this to
```
hdrs = hdr_provider[CcInfo].compilation_context.headers.to_list())
```
but that actually fails, while `direct_headers` works fine!

I'm considering 
```
        hdrs = []

        if CcInfo in hdr_provider:
            hdrs.extend(hdr_provider[CcInfo].compilation_context.headers.to_list())

        if hasattr(hdr_provider, "objc"):
            hdrs.extend(hdr_provider.objc.direct_headers)
```

which should cover all bases, but since there's no case of which we're aware that fails with `direct_headers` alone, I'm not sure...

What do you think @jerrymarino?

File: BazelExtensions/extensions.bzl [hunk 13]
```diff
@@ -324,10 +324,13 @@ def _make_headermap_impl(ctx):
 
     # Extract propagated headermaps
     for hdr_provider in ctx.attr.deps:
-        if not hasattr(hdr_provider, "objc"):
-            continue
+        hdrs = []
 
-        hdrs = hdr_provider.objc.direct_headers
+        if CcInfo in hdr_provider:
+            hdrs.extend(hdr_provider[CcInfo].compilation_context.headers.to_list())
```
Comment (dgcoffman) [inline BazelExtensions/extensions.bzl:330]: Some targets do return the CcInfo provider with an embedded CompilationContext that should have the same headers that were previously attached to ObjcProvider -- but doing this alone fails with error
Screen Shot 2021-03-11 at 10 47 07 AM

**Comment by farcaller** [comment]
I'm trying to build the `React` Exmple app off this branch, and after running `make && bazel build //PodsHost:ios-app` I get random errors on headers missing, e.g.:

```
bazel build //PodsHost:ios-app
INFO: Analyzed target //PodsHost:ios-app (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
ERROR: /Users/farcaller/src/rn-bazel/PodToBUILD/Examples/React/Vendor/React/BUILD.bazel:2006:13: Compiling Vendor/React/React/CxxBridge/RCTCxxBridge.mm failed: (Exit 1): wrapped_clang failed: error executing command external/local_config_cc/wrapped_clang -arch x86_64 '-stdlib=libc++' '-std=gnu++11' '-D_FORTIFY_SOURCE=1' -fstack-protector -fcolor-diagnostics -Wall -Wthread-safety -Wself-assign ... (remaining 81 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox wrapped_clang failed: error executing command external/local_config_cc/wrapped_clang -arch x86_64 '-stdlib=libc++' '-std=gnu++11' '-D_FORTIFY_SOURCE=1' -fstack-protector -fcolor-diagnostics -Wall -Wthread-safety -Wself-assign ... (remaining 81 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
Vendor/React/React/CxxBridge/RCTCxxBridge.mm:19:9: fatal error: 'React/RCTDevSettings.h' file not found
#import <React/RCTDevSettings.h>
        ^~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
Target //PodsHost:ios-app failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 71.352s, Critical Path: 36.85s
INFO: 45 processes: 9 internal, 36 darwin-sandbox.
FAILED: Build did NOT complete successfully
```

```
INFO: Build options --apple_platform_type, --compilation_mode, --cpu, and 3 more have changed, discarding analysis cache.
INFO: Analyzed target //PodsHost:ios-app (34 packages loaded, 14031 targets configured).
INFO: Found 1 target...
ERROR: /Users/farcaller/src/rn-bazel/PodToBUILD/Examples/React/Vendor/React/BUILD.bazel:5344:13: Compiling Vendor/React/React/CoreModules/RCTWebSocketModule.mm failed: (Exit 1): wrapped_clang failed: error executing command external/local_config_cc/wrapped_clang -arch x86_64 '-stdlib=libc++' '-std=gnu++11' '-D_FORTIFY_SOURCE=1' -fstack-protector -fcolor-diagnostics -Wall -Wthread-safety -Wself-assign ... (remaining 85 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox wrapped_clang failed: error executing command external/local_config_cc/wrapped_clang -arch x86_64 '-stdlib=libc++' '-std=gnu++11' '-D_FORTIFY_SOURCE=1' -fstack-protector -fcolor-diagnostics -Wall -Wthread-safety -Wself-assign ... (remaining 85 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
Vendor/React/React/CoreModules/RCTWebSocketModule.mm:14:9: fatal error: 'React/RCTSRWebSocket.h' file not found
#import <React/RCTSRWebSocket.h>
        ^~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
Target //PodsHost:ios-app failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 182.230s, Critical Path: 40.33s
INFO: 265 processes: 136 internal, 129 darwin-sandbox.
FAILED: Build did NOT complete successfully
```

Looks like some include paths didn't propagate properly?

**Comment by dgcoffman** [comment]
@farcaller I think you need the specific BAZEL_OPTS from Examples/React/Makefile. Does `cd Examples/React && make all` work for you?

File: BazelExtensions/extensions.bzl (hunk 12 above)
Comment (jerrymarino) [reply BazelExtensions/extensions.bzl]: @dgcoffman I'll have to pull this down and take a look at to give a clear answer or on it but - adding direct headers at the end seems right. There are likely some edge cases here which may or may not be exhibited in an integration test. If RN was running on the CI and this PR broke it that's probably something to look out for and vet the logic against. The CI is green ✅ 🚢

**Comment by jerrymarino** [comment]
It looks like the GitHub action will build ReactNative and pass if it's working, is that not happening @farcaller ? If the former is happening we should be good to land this - thanks again for the PR @dgcoffman  👍

File: BazelExtensions/extensions.bzl (hunk 13 above)
Comment (jerrymarino) [reply BazelExtensions/extensions.bzl:330]: This makes sense, I think what you've done here and seems more aligned with what we were doing before as opposed to only passing `direct_headers`.  It may have been failing under `darwin-sandbox` if it didn't get all the headers.

**Comment by farcaller** [comment]
Thanks for 47f2d72, makes a bit more sense now!

File: Examples/React/BUILD [hunk 14]
```diff
@@ -1,14 +1,41 @@
...
+        "//Vendor/React",
+        "//Vendor/React:React_hmap",
+    ],
+    copts = [
+        "-I$(GENDIR)/Vendor/React/React_hmap.hmap",
+        "-I."
+    ],
+)
+
+ios_application(
```
Comment (dgcoffman) [inline Examples/React/BUILD:29]: @jerrymarino I expended the React example to actually build the iOS app because there are a couple gotchas in there that took me some time to figure out:
1. The 2016 version of Folly doesn't compile, and the newer v2020.01.13.00 requires an odd workaround
2. You have to use the React_hmap in order for your application code to import from `<React/Whatever>`

**Comment by loyaltyarm** [comment]
Just curious if there is a plan to cut a release with this merged? Thanks in advance!
